kytos/logging/filters.py

```diff
@@ -0,0 +1,123 @@
+
+import time
+from logging import LogRecord
+from threading import Lock
+from typing import Generic, Self, TypeVar
+
+T = TypeVar("T")
+
+
+class _DoublyLinkedList(Generic[T]):
+    value: T
+    previous: Self
+    next: Self
+
+    def __init__(self, value: T):
+        self.value = value
+        self.previous = self
+        self.next = self
+
+    def swap_next(self, other: Self):
+
+        curr_next = self.next
+        other_curr_next = other.next
+
+        self.next = other_curr_next
+        other_curr_next.previous = self
+        other.next = curr_next
+        curr_next.previous = other
+
+    def swap_previous(self, other: Self):
+        curr_prev = self.previous
+        other_curr_prev = other.previous
+
+        self.previous = other_curr_prev
+        other_curr_prev.next = self
+        other.previous = curr_prev
+        curr_prev.next = other
+
+    def remove(self):
+        self.swap_next(self.previous)
+
+    def __iter__(self):
+        yield self.value
+        curr = self.next
+        while curr is not self:
+            yield curr.value
+            curr = curr.next
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class _LimitedCache(Generic[K, V]):
+    root: _DoublyLinkedList[tuple[K, V] | None]
+    cache: dict[K, _DoublyLinkedList[tuple[K, V]]]
+    max_size: int
+
+    def __init__(self, max_size=512):
+        self.root = _DoublyLinkedList(None)
+        self.cache = {}
+        self.max_size = max_size
+
+    def __getitem__(self, key: K) -> V:
+        entry = self.cache[key]
+        _, v = entry.value
+        return v
+
+    def __contains__(self, key: K) -> bool:
+        return key in self.cache
+
+    def __setitem__(self, key: K, value: V):
+        if key in self.cache:
+            entry = self.cache[key]
+            entry.remove()
+            entry.value = (key, value)
+        else:
+            entry = _DoublyLinkedList(
+                (key, value)
+            )
+            self.cache[key] = entry
+            if len(self.cache) == self.max_size:
+                self._remove_oldest()
+
+        self.root.swap_next(entry)
+
+    def __delitem__(self, key: K):
+        entry = self.cache[key]
+        del self.cache[key]
+        entry.remove()
+
+    def _remove_oldest(self):
+        oldest_entry = self.root.previous
+        k, _ = oldest_entry.value
+        del self.cache[k]
+        oldest_entry.remove()
+
+
+class RepeateMessageFilter:
+    lockout_time: float
+    _cache: _LimitedCache
+    _lock: Lock
+
+    def __init__(self, lockout_time: float, cache_size: int = 512):
```

Comment:
512 seems good for now, is the main usage meant to be used only for a single module as you've used with pacing? If future usage is for any module then not sure if 512 is too low, I don't look forward to have to potentially patch this value soon, so if the use is generic per module, let's not overshoot the capacity, but let's not use a too small either.

Reply:
Well, if this ever gets used by other modules in the future, this should instead be a user configurable parameter. If we switch to using dict config for logging, then the user can control the filter, instead of it having to be hard coded in like it is now. From there we can work on recommendations on what value it should be.